kedro/framework/cli/catalog.py

```diff
@@ -73,15 +77,30 @@ def list_datasets(metadata: ProjectMetadata, pipeline, env):
         default_ds = pipeline_ds - catalog_ds
         used_ds = catalog_ds - unused_ds
 
+        # resolve any factory datasets in the pipeline
+        factory_ds_by_type = defaultdict(list)
+        for ds_name in default_ds:
+            matched_pattern = data_catalog._match_pattern(
+                data_catalog._dataset_patterns, ds_name
+            )
+            if matched_pattern:
+                ds_config = data_catalog._resolve_config(ds_name, matched_pattern)
+                factory_ds_by_type[ds_config["type"]].append(ds_name)
+
+        default_ds = default_ds - set(chain.from_iterable(factory_ds_by_type.values()))
```

Comment:
What does `chain.from_iterable()` do here?

Reply:
Here `chain.from_iterable()` takes the collection of lists from `factory_ds_by_type.values()` (as the value type of this dict is lists) and flattens them into a single, flattened iterable.